inst/include/dust/dust.hpp

```diff
@@ -45,8 +45,93 @@ struct pars_t {
     pars_t(dust::nothing(), internal_) {
   }
 };
-}
 
+// If we store history, we need to save:
+//
+// history order (n_particles * n_data)
+// history value (n_history_state * n_particles * n_data)
+//
+// restart (not yet thinking about this)
+//
+// implies that we should save an index
+
+template <typename real_t>
+class filter_state {
+public:
+  filter_state(size_t n_state, size_t n_particles, size_t n_data) :
+    n_state_(n_state), n_particles_(n_particles), n_data_(n_data), offset_(0) {
+    resize(n_state, n_particles, n_data);
+  }
+
+  // default constructable
+  filter_state() : filter_state(0, 0, 0) {
+  }
+
+  void resize(size_t n_state, size_t n_particles, size_t n_data) {
```

Comment:
This doesn't reset, so presume it's if you want to increase the size. Can the size shrink here too?

Reply:
Looks like the class is initialised with the default constructor, and then resize used to grow this if data is provided

Reply:
So do we always create the class, and then set the data on it? Is there an argument for setting the data when the class is initialised to simplify these constructors? Or is that flexibility needed?

Reply:
There's an argument for that, but we'd probably still retain this at the low level to keep the contructors simple and do that initialisation in R

Reply:
> This doesn't reset, so presume it's if you want to increase the size. Can the size shrink here too?

Size can shrink yes, though the memory won't be reclaimed or cause a copy when that happens

> Looks like the class is initialised with the default constructor, and then resize used to grow this if data is provided

exactly